File: 10-heap/00973-K_Closest_Points_To_Origin/00973-k_closest_points_to_origin.cpp
// 973: K Closest Points To Origin
// https://leetcode.com/problems/k-closest-points-to-origin


#include <iostream>
#include <queue>
#include <vector>

using namespace std;


class Solution {
public:
    // SOLUTION
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        vector<vector<int>> triples;
        for (auto& p : points)
            triples.push_back({p[0] * p[0] + p[1] * p[1], p[0], p[1]});

        priority_queue<vector<int>, vector<vector<int>>, greater<vector<int>>> pq(triples.begin(), triples.end());
        vector<vector<int>> result;

        while (k--){
            vector<int> e = pq.top();
            pq.pop();
            result.push_back({e[1], e[2]});
        }

        return result;
    }
};


int main() {
    Solution o;

    // INPUT
    vector<vector<int>> points = {{1,3},{-2,2}};
    int k = 1;

    // OUTPUT
    auto result = o.kClosest(points, k);

    cout << "["; for (auto x : result) {
        cout << "["; for (auto y : x) {
            cout << y << ",";
        } cout << "\b],";
    } cout << "\b]" << endl;

    return 0;
}

File: 10-heap/00973-K_Closest_Points_To_Origin/00973-k_closest_points_to_origin.java
// 973: K Closest Points To Origin
// https://leetcode.com/problems/k-closest-points-to-origin


import java.util.PriorityQueue;


class Solution {
    // SOLUTION
    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) ->
            Integer.compare(
                (a[0] * a[0] + a[1] * a[1]),
                (b[0] * b[0] + b[1] * b[1])
            )
        );

        for (int[] point : points) {
            pq.add(point);
        }

        int[][] result = new int[k][2];
        for (int i = 0; i < k; i++) {
            int[] e = pq.poll();
            result[i][0] = e[0];
            result[i][1] = e[1];
        }

        return result;
    }


    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        int[][] points = {{1,3},{-2,2}};
        int k = 1;

        // OUTPUT
        var result = o.kClosest(points, k);

        System.out.print("["); for (var x : result) {
            System.out.print("["); for (var y : x) {
                System.out.print(y + ",");
            } System.out.print("\b],");
        } System.out.println("\b]");
    }
}
